src/riscv/lib/src/state_backend/proof_layout.rs

```diff
@@ -275,17 +275,30 @@ impl OwnedProofPart {
         }
     }
 
-    /// Map the inner [`MerkleProof`] if the proof is present.
-    ///
-    /// Panic: In debug mode, panic if the proof is absent.
-    pub(crate) fn map_present_unchecked(self, f: impl FnOnce(MerkleProof)) {
-        match self {
-            OwnedProofPart::Absent => debug_assert!(
-                false,
-                "If the node is present, then the branch is also present"
-            ),
-            OwnedProofPart::Present(proof) => f(proof),
+    /// Construct a node from its child proofs. The `parent` parameter allows us to restruct the
+    /// blinded state of the parent.
+    pub fn node_from_children(
+        parent: Partial<()>,
+        children: impl IntoIterator<Item = Self>,
+    ) -> Self {
+        match parent {
+            Partial::Absent => return OwnedProofPart::Absent,
+            Partial::Blinded(hash) => {
+                return OwnedProofPart::Present(MerkleProof::leaf_blind(hash));
+            }
+            Partial::Present(_) => {}
+        }
+
+        let mut partial_children = Vec::with_capacity(MERKLE_ARITY);
+
+        for item in children {
+            match item {
+                OwnedProofPart::Absent => return OwnedProofPart::Absent,
```

Comment:
to double check - if _any_ children are absent, we mark the whole node as absent?

Should it be if all the children are absent, then the node is absent? If only one out of two children are absent, should that instead be a logic error?

Reply:
apologies if I've misunderstood something though

Reply:
It is not possible to represent a present node with partially absent children. The proof representation does not allow it. If any of the children are absent, then the proof for that node is also absent.

The closest analogy I can think of is constructing (or transposing in this case) an `Option<Vec<T>>` from `Vec<Option<T>>` where any `None` element will result in a global `None`.